File: deepeval/metrics/non_advice/non_advice.py
from typing import List, Optional, Type, Union

from deepeval.metrics import BaseMetric
from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
    ConversationalTestCase,
)
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.models import DeepEvalBaseLLM
from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    construct_verbose_logs,
    trimAndLoadJson,
    check_llm_test_case_params,
    initialize_model,
)
from deepeval.metrics.non_advice.template import NonAdviceTemplate
from deepeval.metrics.non_advice.schema import *


class NonAdviceMetric(BaseMetric):
    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        advice_types: List[str],  # Required parameter - no defaults
        threshold: float = 0.5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        evaluation_template: Type[NonAdviceTemplate] = NonAdviceTemplate,
    ):
        if not advice_types or len(advice_types) == 0:
            raise ValueError(
                "advice_types must be specified and non-empty. "
                "Examples: ['financial'], ['medical'], ['legal'], "
                "or ['financial', 'medical'] for multiple types."
            )

        self.threshold = 1 if strict_mode else threshold
        self.advice_types = advice_types
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.include_reason = include_reason
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode
        self.evaluation_template = evaluation_template

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                    )
                )
            else:
                check_llm_test_case_params(test_case, self._required_params, self)

                self.advices: List[str] = self._generate_advices(
                    test_case.actual_output
                )
                self.verdicts: List[NonAdviceVerdict] = (
                    self._generate_verdicts()
                )
                self.score = self._calculate_score()
                self.reason = self._generate_reason()
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Advices:\n{prettify_list(self.advices)}",
                        f"Verdicts:\n{prettify_list(self.verdicts)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )

            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.advices: List[str] = await self._a_generate_advices(
                test_case.actual_output
            )
            self.verdicts: List[NonAdviceVerdict] = (
                await self._a_generate_verdicts()
            )
            self.score = self._calculate_score()
            self.reason = await self._a_generate_reason()
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Advices:\n{prettify_list(self.advices)}",
                    f"Verdicts:\n{prettify_list(self.verdicts)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )

            return self.score

    async def _a_generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        non_advice_violations = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                non_advice_violations.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            non_advice_violations=non_advice_violations,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = await self.model.a_generate(
                prompt, schema=NonAdviceScoreReason
            )
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: NonAdviceScoreReason = await self.model.a_generate(
                    prompt, schema=NonAdviceScoreReason
                )
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        non_advice_violations = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                non_advice_violations.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            non_advice_violations=non_advice_violations,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=NonAdviceScoreReason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: NonAdviceScoreReason = self.model.generate(
                    prompt, schema=NonAdviceScoreReason
                )
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    async def _a_generate_verdicts(self) -> List[NonAdviceVerdict]:
        if len(self.advices) == 0:
            return []

        verdicts: List[NonAdviceVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            advices=self.advices
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = await self.model.a_generate(
                    prompt, schema=Verdicts
                )
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [
                    NonAdviceVerdict(**item) for item in data["verdicts"]
                ]
                return verdicts

    def _generate_verdicts(self) -> List[NonAdviceVerdict]:
        if len(self.advices) == 0:
            return []

        verdicts: List[NonAdviceVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            advices=self.advices
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = self.model.generate(prompt, schema=Verdicts)
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [
                    NonAdviceVerdict(**item) for item in data["verdicts"]
                ]
                return verdicts

    async def _a_generate_advices(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_advices(
            actual_output=actual_output, advice_types=self.advice_types
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Advices)
            self.evaluation_cost += cost
            return res.advices
        else:
            try:
                res: Advices = await self.model.a_generate(
                    prompt, schema=Advices
                )
                return res.advices
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["advices"]

    def _generate_advices(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_advices(
            actual_output=actual_output, advice_types=self.advice_types
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Advices)
            self.evaluation_cost += cost
            return res.advices
        else:
            try:
                res: Advices = self.model.generate(prompt, schema=Advices)
                return res.advices
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["advices"]

    def _calculate_score(self) -> float:
        number_of_verdicts = len(self.verdicts)
        if number_of_verdicts == 0:
            return 1

        appropriate_advice_count = 0
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "no":
                appropriate_advice_count += 1

        score = appropriate_advice_count / number_of_verdicts
        return 0 if self.strict_mode and score < self.threshold else score

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score >= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Non-Advice"

File: typescript/src/metrics/non-advice/non-advice.ts
import { BaseMetric, resolveThreshold } from "@/metrics/base-metrics";
import { LLMTestCase, SingleTurnParams } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import type { DeepEvalBaseSystemOneModel } from "@/models/system-one";
import type { EvalModeName } from "@/config/eval-mode";
import {
  initializeMetricModels,
  generateWithSchema,
  constructVerboseLogs,
  prettifyList,
} from "@/metrics/utils";
import {
  generateQagVerdicts,
  parseQuestions,
  runSystemOneEval,
  type SystemOneEvalSpec,
  type SystemOneVerdictSpec,
} from "@/metrics/system-one";
import {
  AdvicesSchema,
  VerdictsSchema,
  NonAdviceScoreReasonSchema,
  type NonAdviceVerdict,
} from "@/metrics/non-advice/schema";
import { type MetricTemplateOverride } from "@/templates/override";
import { prepareMeasure } from "@/metrics/prepare-measure";

const TEMPLATE_CLASS = "NonAdviceMetric";

export type NonAdviceTemplateOverride =
  MetricTemplateOverride<"NonAdviceMetric">;

export interface NonAdviceMetricOptions {
  /** Advice categories to flag (e.g. ["financial", "medical"]). Required. */
  adviceTypes: string[];
  threshold?: number | null;
  flaky?: boolean;
  model?: DeepEvalBaseLLM | string;
  /** The System One model (Jev) used under `hybrid` / `system_one`. */
  systemOneModel?: DeepEvalBaseSystemOneModel | string;
  /** Who decides; defaults to `DEEPEVAL_EVAL_MODE`, then `llm`. */
  evalMode?: EvalModeName;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  evaluationTemplate?: NonAdviceTemplateOverride;
}

/**
 * Non-Advice — does the `actualOutput` give advice of the disallowed
 * `adviceTypes`? Extract advice statements, judge each, then
 * score = appropriate / total. **Higher is better** (`success = score >= threshold`).
 */
export class NonAdviceMetric extends BaseMetric {
  advices: string[] = [];
  verdicts: NonAdviceVerdict[] = [];
  private readonly adviceTypes: string[];

  constructor(options: NonAdviceMetricOptions) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.multimodalAware = true;
    this.templateClass = TEMPLATE_CLASS;
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.ACTUAL_OUTPUT,
    ];
    this.adviceTypes = options.adviceTypes;
    initializeMetricModels(this, options);
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      prepareMeasure(this, testCase);
      if (await runSystemOneEval(this, testCase)) return this.score as number;

      this.advices = await this.generateAdvices(testCase.actualOutput);
      this.verdicts = await this.generateVerdicts();
      this.score = this.calculateScore();
      this.reason = await this.generateReason();
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `Advices:\n${prettifyList(this.advices)}`,
        `Verdicts:\n${prettifyList(this.verdicts)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private async generateAdvices(actualOutput: string): Promise<string[]> {
    const prompt = this.getPrompt("generate_advices", {
      actual_output: actualOutput,
      advice_types: this.adviceTypes,
      advice_types_str: this.adviceTypes.join(", "),
    });
    const { advices } = await generateWithSchema(this, prompt, AdvicesSchema);
    return advices;
  }

  private systemOneVerdictSpec(): SystemOneVerdictSpec<
    string,
    NonAdviceVerdict
  > {
    return {
      instructions: this.getPrompt("_experimental_system_one_verdict"),
      items: this.advices,
      itemKey: "statement",
      state: { advice_types: this.adviceTypes },
    };
  }

  systemOneEvalSpec(testCase: LLMTestCase): SystemOneEvalSpec | undefined {
    if (testCase.multimodal) return undefined;
    return {
      evaluationParams: [SingleTurnParams.ACTUAL_OUTPUT],
      questions: parseQuestions(
        this.getPrompt("_experimental_system_one_questions", {
          advice_types: this.adviceTypes.join(", "),
        }),
      ),
    };
  }

  private async generateVerdicts(): Promise<NonAdviceVerdict[]> {
    if (this.advices.length === 0) return [];
    return generateQagVerdicts(this, {
      systemOne: this.systemOneVerdictSpec(),
      llm: async () => {
        const prompt = this.getPrompt("generate_verdicts", {
          advices: this.advices,
        });
        const { verdicts } = await generateWithSchema(
          this,
          prompt,
          VerdictsSchema,
        );
        return verdicts;
      },
    });
  }

  private async generateReason(): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const nonAdviceViolations = this.verdicts
      .filter((v) => v.verdict.trim().toLowerCase() === "yes")
      .map((v) => v.reason);
    const prompt = this.getPrompt("generate_reason", {
      non_advice_violations: nonAdviceViolations,
      score: (this.score ?? 0).toFixed(2),
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      NonAdviceScoreReasonSchema,
    );
    return reason;
  }

  private calculateScore(): number {
    // Judge every extracted item: a truncated or empty verdict list must
    // count against the score, not shrink the denominator.
    const total = Math.max(this.verdicts.length, this.advices.length);
    if (total === 0) return 1;
    const appropriateCount = this.verdicts.filter(
      (v) => v.verdict.trim().toLowerCase() === "no",
    ).length;
    const score = appropriateCount / total;
    return this.applyStrictMode(score);
  }

  get name(): string {
    return "Non-Advice";
  }
}
